[build-utils] add default shouldServe and type ShouldServeOptions

amio opened the pull request:
Added the `ShouldServeParam` type declaration and default implementation for `shouldServe` for builder maker to import.

styfle commented on packages/now-build-utils/src/types.ts:
```diff
@@ -99,3 +99,26 @@ export interface PrepareCacheOptions {
+   * An arbitrary object passed by the user in the build definition defined
+   * in `now.json`.
+   */
+  config?: object;
```
```suggestion
  config: Config;
```

Should this be the same config from the `BuildOptions`?

amio replied:
Yes it is 👍
but it should be optional

styfle replied:
@amio Why should it be optional?

We intentionally made the config object always exists before calling the builder.

https://github.com/zeit/lambdas/pull/368

amio replied:
In now-dev's case it's omitted 🤔
https://github.com/zeit/now-cli/blob/0f74a0695c6de3605ebca616ec373c9cc7c7cca1/src/commands/dev/lib/dev-server.ts#L680

styfle replied:
@amio Ok then let's remove line 123 because it looks like there is never a config passed in.

amio replied:
I guess so, but not sure if @TooTallNate had a plan for it

amio replied:
It's listed here, in `static-builder`'s `ShouldServeParams`: https://github.com/zeit/now-cli/blob/0f74a0695c6de3605ebca616ec373c9cc7c7cca1/src/commands/dev/lib/types.ts#L115

styfle replied:
@amio I just deleted that property from the `now-cli` and everything built fine. I would leave it off because we can always add it later in a minor update but we can never remove it without a breaking change.

styfle commented on packages/now-node/package.json:
```diff
@@ -22,6 +22,7 @@
     "dist"
   ],
   "devDependencies": {
+    "@now/build-utils": "^0.4.40",
```
The `@now/build-utils` are not normally included in the builder devDependencies. Please remove.

amio replied:
Why does build task keep complaining about this?

```
/home/circleci/repo/packages/now-node/test/test.js
   4:20  error  '@now/build-utils' should be listed in the project's dependencies, not devDependencies  import/no-extraneous-dependencies
  38:1   error  'test' is not defined                                                                   no-undef

✖ 2 problems (2 errors, 0 warnings)
```

amio replied:
@styfle Found we can eliminate it by this https://github.com/zeit/now-builders/pull/382/commits/d2166d8df9848aef1c85bc1d03dbccb7b85ecc59 😌

styfle commented on packages/now-node/src/index.ts:
```diff
@@ -127,3 +129,29 @@ export async function prepareCache({ workPath }: PrepareCacheOptions) {
   };
 }
+
+export function shouldServe({
```
Is this what we are considering the "default" shouldServe?

If that's the case, I think we should move this implementation from this builder to `build-utils`.

amio replied:
Brilliant! cc @TooTallNate

amio commented on packages/now-node/src/index.ts:
```diff
@@ -127,3 +129,29 @@ export async function prepareCache({ workPath }: PrepareCacheOptions) {
+  entrypoint,
+  files,
+  requestPath
+}: ShouldServeParams): boolean {
```
@TooTallNate I slightly tweaked this method, you may take look at if I missed any possible cases.

TooTallNate commented on packages/now-build-utils/src/types.ts:
```diff
@@ -99,3 +99,21 @@ export interface PrepareCacheOptions {
   config: Config;
 }
+
+export interface ShouldServeParams {
```
`config` should definitely be in here. It's a bug that `now dev` doesn't pass it in so I'll adjust that.

styfle replied:
@TooTallNate Can we assume that `config` is never undefined like we do for `build()`?

amio commented:
@styfle @TooTallNate Update:

- Removed `shouldServe` implementation in favor of https://github.com/zeit/now-cli/pull/2150
- Added back the `config` param (optional)

Ready to review now :D

styfle commented on packages/now-node/src/index.ts:
```diff
@@ -1,5 +1,5 @@
-import { remove, readFile } from 'fs-extra';
+import { readFile } from 'fs-extra';
 import {
   glob,
```
@amio Actually, you need to add `ShouldServeParams` here.

amio replied:
Why here? I think it's for all builders

styfle replied:
@amio Sorry, I saw `index.ts` and didn't look at the path. What I mean is that type needs to exported from `now-build-utils/src/index.ts`

styfle commented on packages/now-build-utils/src/types.ts:
```diff
@@ -99,3 +99,26 @@ export interface PrepareCacheOptions {
   config: Config;
 }
+
+export interface ShouldServeParams {
```
```suggestion
export interface ShouldServeOptions {
```

Can we make this match the others?

amio replied:
Thought about that, but I felt it's more like **params** than **options** to  `shouldServe`. "options" might more suits to some function with side effect. 🤔

styfle replied:
I made a bad decision when I added TypeSCript and used `AnalyzeOptions` `BuildOptions`, and `PrepareCacheOptions` to name these inputs. These are already documented here: https://zeit.co/docs/v2/deployments/builders/developer-guide/#builder-exports

I think we need to maintain the same naming convention when introducing to params/options.

styfle commented on packages/now-build-utils/src/types.ts:
```diff
@@ -99,3 +99,26 @@ export interface PrepareCacheOptions {
+   * An arbitrary object passed by the user in the build definition defined
+   * in `now.json`.
+   */
+  config?: Config;
```
```suggestion
  config: Config;
```

As discussed in slack, lets make this never undefined

amio replied:
IMO this should keep what it is from user's config, it is optional in that case. It would be strange for builder runtime to get a `{}` or something else when the builder config has no `config` attr.

styfle replied:
Why is that strange? It allows the builder to stop checking for existence which prevents duplication in each builder and should also improve performance a bit.

amio replied:
Because we will tell user this is the `config` from builder config in `now.json`.

While we can add that "it's default value is `{}`".

styfle replied:
We already do this for the `build()` function.

Can you think of a use case where a builder would need to separate `undefined` vs `{}`?

amio replied:
Just prefer the way needs less explaining ☺️

amio replied:
While I'm not strongly against this either :D

amio commented:
@styfle @TooTallNate Updated this PR:

- Added a default `shouldServe` implementation
- Make `config` required

styfle commented on packages/now-build-utils/src/should-serve.ts:
```diff
@@ -0,0 +1,26 @@
+}
+
+function hasProp (obj: Map<string, any>, key: string): boolean {
+  return Object.hasOwnProperty.call(obj, key)
```
```suggestion
  return obj.has(key);
```

If this is truly a map, you can use has method. But then you don't really need `hasProp` helper at all 🤷‍♂️

styfle commented on packages/now-build-utils/src/types.ts:
```diff
@@ -99,3 +101,26 @@ export interface PrepareCacheOptions {
+  /**
+   * All source files of the project
+   */
+  files: Map<string, FileFsRef>;
```
Are you sure this is a Map? Where does it get defined in now-cli?

amio replied:
Yep, it should be an object, fixed

amio replied:
Yep, it should be an object literal, fixed

styfle commented on packages/now-build-utils/src/should-serve.ts:
```diff
@@ -0,0 +1,26 @@
+  return false;
+}
+
+function hasProp (obj: Map<string, any>, key: string): boolean {
```
```suggestion
function hasProp(obj: { [path: string]: FileFsRef }, key: string): boolean {
```
